### PR TITLE
use Atlas structures with an IFS-like variable batching

### PR DESCRIPTION
**Pull request by sbrdar**
Using Atlas structures in CLOUDSC via configuration flag "--with-atlas"
1) variable batching (as in GML fields of IFS) is enabled by default via atlas::MultiField
2) no batching is possible by setting the environment variable CLOUDSC_ATLAS_MULTIFIELD to 0
* with Willem Deconinck
* thanks to Balthasar Reuter

### CODE REVIEW

**Comment by reuterbal** [comment]
With #54 merged, this needs a rebase. @sbrdar Can you take care of that so we can bring this variant in as well?

**Comment by sbrdar** [comment]
> With #54 merged, this needs a rebase. @sbrdar Can you take care of that so we can bring this variant in as well?

I have merged now 'develop' into 'nab_atlas_MultiField'.

File: bundle.yml [hunk 1]
```diff
@@ -61,7 +61,7 @@ projects :
 
     - atlas :
         git     : https://github.com/ecmwf/atlas
-        version : 0.34.0
+        version : feature/MultiField
```
Comment (reuterbal) [inline bundle.yml:64]: Just out of curiosity: Is there a timeline for including this feature in a release?

File: src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90 [hunk 2]
```diff
@@ -129,11 +127,15 @@ MODULE CLOUDSC_GLOBAL_ATLAS_STATE_MOD
...
     CLASS(CLOUDSC_GLOBAL_ATLAS_STATE_BLOCK_VIEW), INTENT(INOUT) :: SELF
-    CLASS(ATLAS_FIELDSET), INTENT(INOUT) :: FSET
+    TYPE(ATLAS_FIELDSET), INTENT(INOUT) :: FSET
     INTEGER, INTENT(IN) :: IBLK
 
     REAL(KIND=JPRB), POINTER :: TMP3D(:,:,:)
 
+!    CALL FSET%UPDATE_DEVICE(1,IBLK)
+!    field = fset%field(1)
+!    call field%update_device()
```
Comment (reuterbal) [inline src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:137]: Debug leftover?

File: src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90 [hunk 3]
```diff
@@ -190,73 +192,136 @@ SUBROUTINE CLOUDSC_GLOBAL_ATLAS_STATE_BLOCK(SELF, FSET, IBLK)
...
+
+
+    OUTVAR_SIZE = SIZE(OUT_VAR_NAMES) - 2
+    IF (LMULTIFIELD) THEN
+        ! output multifield on model interfaces, i.e. LEVELS = FSPACE%LEVELS() + 1 = SELF%KLEV + 1
+        DO IVAR = 1, OUTVAR_SIZE
+            OUT_MFIELD_CONFIG(IVAR) = ATLAS_CONFIG()
+            CALL OUT_MFIELD_CONFIG(IVAR)%SET("name", TRIM(OUT_VAR_NAMES(IVAR)))
+            CALL OUT_MFIELD_CONFIG(IVAR)%SET("levels", SELF%KLEV+1)
+        !    CALL FSET%ADD(FSPACE%CREATE_FIELD(NAME=TRIM(OUT_VAR_NAMES(IVAR)), KIND=ATLAS_REAL(JPRB)))
```
Comment (reuterbal) [inline src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:298]: debug leftover?

File: src/cloudsc_fortran_atlas/validate_atlas_mod.F90 [hunk 4]
```diff
@@ -59,18 +60,17 @@ SUBROUTINE VALIDATEVAR_ATLAS(FSET, NAME, NLON, NGPTOTG, STATE_VAR)
...
     NBLOCKS = FSPACE%NBLKS()
-    !NPROMA = FSPACE%NPROMA()
     NPROMA = FIELD%SHAPE(1)
 
     ZMINVAL(1) = +HUGE(ZMINVAL(1))
     ZMAX_VAL_ERR(1) = -HUGE(ZMAX_VAL_ERR(1))
     ZMAX_VAL_ERR(2) = 0.0_JPRB
     ZSUM_ERR_ABS(:) = 0.0_JPRB
 
+    !CALL INPUT_INITIALIZE(NAME='reference')
```
Comment (reuterbal) [inline src/cloudsc_fortran_atlas/validate_atlas_mod.F90:73]: Debug leftover?

File: src/cloudsc_fortran_atlas/validate_atlas_mod.F90 [hunk 5]
```diff
@@ -160,6 +160,7 @@ SUBROUTINE VALIDATEVAR_ATLAS(FSET, NAME, NLON, NGPTOTG, STATE_VAR)
         PRINT *, "FIELD RANK NOT SUPPORTED"
         CALL EXIT(1)
     ENDIF
+    !CALL INPUT_FINALIZE()
```
Comment (reuterbal) [inline src/cloudsc_fortran_atlas/validate_atlas_mod.F90:163]: Debug leftover?

File: src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90 [hunk 6]
```diff
@@ -190,73 +192,136 @@ SUBROUTINE CLOUDSC_GLOBAL_ATLAS_STATE_BLOCK(SELF, FSET, IBLK)
...
+    TRACE = ATLAS_TRACE("cloudsc_global_atlas_state_mod.F90", __LINE__, &
+        & "CLOUDSC_GLOBAL_ATLAS_STATE_LOAD")
+
+    LMULTIFIELD = .TRUE.
+    CALL GET_ENVIRONMENT_VARIABLE("CLOUDSC_ATLAS_MULTIFIELD",CENV,CENV_LEN)
+    IF (CENV_LEN > 0 ) THEN
+      IF( TRIM(CENV) == "0" .OR. TRIM(CENV) == "OFF" .OR. TRIM(CENV) == "FALSE" ) THEN
+        LMULTIFIELD = .FALSE.
+      ENDIF
+    ENDIF
```
Comment (reuterbal) [inline src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:234]: I just noticed that we don't have a description of this variant in the README. Could you add just 2-3 sentences that explain what this variant is (e.g., "uses Atlas as storage backend for field data") and that two choices for the backend exist Multifield/FieldSet, which can be switched using that environment variable.

**Review by reuterbal — CHANGES_REQUESTED** [review]
Thanks and apologies for the review latency. This looks great!

I also like the introduction of atlas_trace.

I've flagged a few places with commented code leftovers, which should be removed.

Second request is to add a brief (!) description to the README (just an entry under "prototypes available").

File: bundle.yml (hunk 1 above)
Comment (sbrdar) [reply bundle.yml:64]: Willem is going to include the feature in the release very soon.

File: src/cloudsc_fortran_atlas/validate_atlas_mod.F90 (hunk 4 above)
Comment (sbrdar) [reply src/cloudsc_fortran_atlas/validate_atlas_mod.F90]: Yes. The line is removed.

File: src/cloudsc_fortran_atlas/validate_atlas_mod.F90 (hunk 5 above)
Comment (sbrdar) [reply src/cloudsc_fortran_atlas/validate_atlas_mod.F90]: True. I have removed the line.

File: src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90 (hunk 2 above)
Comment (sbrdar) [reply src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90]: Yes. The lines are removed.

File: src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90 (hunk 6 above)
Comment (sbrdar) [reply src/cloudsc_fortran_atlas/cloudsc_global_atlas_state_mod.F90:230]: Good point! Added 2-3 sentences to README. Thanks for the review!

**Review by reuterbal — APPROVED** [review]
Many thanks for making these changes! Good to go!

**Pull request merged**